x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheService.java

```diff
@@ -0,0 +1,414 @@
+                    success = true;
+                    return cacheEntry;
+                } catch (IOException e) {
+                    logger.error(() -> new ParameterizedMessage("failed to create cache file [{}]", path), e);
+                    throw e;
+                } finally {
+                    if (success == false) {
+                        IOUtils.deleteFilesIgnoringExceptions(path);
+                    }
+                }
+            });
+        } catch (ExecutionException e) {
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+    }
+
+    /**
+     * Reads bytes from a cached file.
+     * <p>
+     * The cached file is identified by its path ({@code filePath}). This path is used as a cache key by the cache service to know if an
+     * entry already exists in cache for the given file.
+     * <p>
+     * When no entry exists in cache, the service creates a new entry for the file. Because cached files are created as sparse files on
+     * disk, each cache entry needs to track the ranges of bytes that have already been cached and the ranges of bytes that will need to be
+     * retrieved from a source. This is tracked in {@link CacheEntry} by {@link SparseFileTracker} which requires to know the length of the
+     * file to cache to correctly track ranges.
+     * <p>
+     * When an entry already exists in cache, the service uses the {@link SparseFileTracker} of the {@link CacheEntry} to know if it can
+     * serve the read operations using one or more ranges of bytes already cached on local disk or if it needs to retrieve the bytes from
+     * the source (supplied as a {@link IndexInput}) first, then writes the bytes locally before returning them as the result of the read
+     * operation.
+     *
+     * @param filePath   the {@link Path} of the cached file
+     * @param fileLength the length of the cached file (required to compute ranges of bytes)
+     * @param fileSource supplies the {@link IndexInput} to read the bytes from in case they are not already in cache
+     * @param position   the position in the cached file where to start reading bytes from
+     * @param b          the array to read bytes into
+     * @param off        the offset in the array to start storing bytes
+     * @param len        the number of bytes to read
+     *
+     * @throws IOException if something went wrong
+     */
+    public void readFromCache(final Path filePath, final long fileLength, final CheckedSupplier<IndexInput, IOException> fileSource,
+                              final long position, final byte[] b, int off, int len) throws IOException {
+        long pos = position;
+        while (len > 0 && lifecycleState() == Lifecycle.State.STARTED) {
+            final CacheEntry cacheEntry = getOrAddToCache(filePath, fileLength, fileSource);
+            if (cacheEntry.tryIncRef() == false) {
+                continue;
```

Comment:
I'm concerned about thrashing here. We're trying to handle the case where the entry is evicted between `getOrAddToCache` and `tryIncRef`, but if we have multiple threads all adding entries that get immediately evicted (might happen if the cache is too small and/or almost full) I think we could end up stuck here. Maybe it'd be better to give up on the cache in this case and read directly from the `fileSource`?

Reply:
You're right, and this is a good suggestion.

I changed the test so that it [randomly uses a smaller cache](https://github.com/elastic/elasticsearch/pull/50693/files#diff-c5c11300911596f2e53b53f8c0d4daedR60) and changed the readFromCache() method so t[hat it serves the read operation directly from the source](https://github.com/elastic/elasticsearch/pull/50693/files#diff-a0405cffc7e6de74715aa3935ac736ecR159-R165) if the picked up cache file is already evicted. It does a [similar thing](https://github.com/elastic/elasticsearch/pull/50693/files#diff-a0405cffc7e6de74715aa3935ac736ecR237-R239) later in the process, if the cache file has been evicted just before writing the new range on disk.